owsi-core/owsi-core-components/owsi-core-component-jpa-more/src/main/java/fr/openwide/core/jpa/more/business/history/service/AbstractHistoryLogServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+package fr.openwide.core.jpa.more.business.history.service;
+
+import java.util.Date;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.google.common.base.Supplier;
+
+import fr.openwide.core.jpa.business.generic.service.GenericEntityServiceImpl;
+import fr.openwide.core.jpa.exception.SecurityServiceException;
+import fr.openwide.core.jpa.exception.ServiceException;
+import fr.openwide.core.jpa.more.business.difference.service.IDifferenceService;
+import fr.openwide.core.jpa.more.business.difference.util.IDifferenceFromReferenceGenerator;
+import fr.openwide.core.jpa.more.business.difference.util.IHistoryDifferenceGenerator;
+import fr.openwide.core.jpa.more.business.history.dao.IAbstractHistoryLogDao;
+import fr.openwide.core.jpa.more.business.history.model.AbstractHistoryDifference;
+import fr.openwide.core.jpa.more.business.history.model.AbstractHistoryLog;
+import fr.openwide.core.jpa.more.business.history.model.bean.AbstractHistoryLogObjectsBean;
+import fr.openwide.core.jpa.more.business.history.util.HistoryLogBeforeCommitTask;
+import fr.openwide.core.jpa.more.business.history.util.HistoryLogBeforeCommitWithDifferencesTask;
+import fr.openwide.core.jpa.more.business.history.util.IDifferenceHandler;
+import fr.openwide.core.jpa.more.util.transaction.service.ITransactionSynchronizationTaskManagerService;
+
+public abstract class AbstractHistoryLogServiceImpl<HL extends AbstractHistoryLog<HL, HET, HD>,
+				HET extends Enum<HET>,
+				HD extends AbstractHistoryDifference<HD, HL>>
+		extends GenericEntityServiceImpl<Long, HL> implements IAbstractHistoryLogService<HL, HET, HD> {
+
+	@Autowired
+	protected ITransactionSynchronizationTaskManagerService transactionSynchronizationTaskManagerService;
+
+	@Autowired
+	protected IHistoryValueService valueService;
+	
+	@Autowired
+	public AbstractHistoryLogServiceImpl(IAbstractHistoryLogDao<HL, HET> dao) {
+		super(dao);
+	}
+	
+	@Override
+	public HL logNow(Date date, HET eventType, List<HD> differences, AbstractHistoryLogObjectsBean<?> objects)
+			throws ServiceException, SecurityServiceException {
+		HL log = newHistoryLog(date, eventType, differences, objects);
+
+		log.setDifferences(differences);
+		for (HD difference : differences) {
+			difference.setParentLog(log);
+		}
+		
+		create(log);
+		return log;
+	}
+	
+	protected abstract HL newHistoryLog(Date date, HET eventType, List<HD> differences, AbstractHistoryLogObjectsBean<?> objects);
+
+	protected abstract Supplier<HD> newHistoryDifferenceSupplier();
+
+	protected void setSecondaryObjects(HL log, AbstractHistoryLogObjectsBean<?> objects) {
```

Comment:
I think we should keep it but find a place for a comment to explain when to use it.